### ISSUE TITLE
AsyncCommand::invoke_script doesn't handle NoScriptError

### ISSUE THREAD

**Issue by svix-jplatte** [mitsuhiko/redis-rs#1394]
`ScriptInvocation::{invoke, invoke_async}` have logic to handle `NoScriptError`, invoking `SCRIPT LOAD` to make the script available and retrying the `EVALSHA` command. However, `AsyncCommands::invoke_script` doesn't do the same. I think it should.

**Comment by svix-jplatte** [comment]
Honestly, I don't know enough about redis to say how that would ideally behave. But one thing is worth noting: I didn't run into this when using a `Pipeline`, it was just a simple connection. 

cc @svix-james

**Comment by nihohit** [comment]
> I didn't run into this when using a Pipeline, it was just a simple connection.

I'm sorry, I didn't understand your intention here. You meant that you deduced that there might be an issue, you didn't experience it in the wild?

> how about documenting the current behavior?

Sounds great. Would you like to open a PR?

**Comment by svix-jplatte** [comment]
No, I did run into this for real. But it wasn't with a `Pipeline`, it was with a `MultiplexedConnection` (both implementing `AsyncCommands`).

**Comment by jaymell** [comment]
Honestly, `invoke_script` was probably not a good name for this command. We should probably just have `eval` and `evalsha` separately, which are much more discoverable given that they are the actual commands used. 

The functionality you're pointing out is documented but should probably be noted on the command itself: https://github.com/redis-rs/redis-rs/blob/main/redis/src/lib.rs#L389

**Comment by sondrelg** [comment]
I did stumble onto this using a pipeline. Given the current design, would this be the way to best handle things?

```rust
pipeline.invoke_script(MY_SCRIPT
    .key(key1)
    .key(key2)
    .arg(&arg1)
    .arg(&arg2)
match pipeline.query_async::<()>(&mut connection)
    .await {
    Ok(_) => {}
    Err(e) => {
        match e.kind() {
            ErrorKind::NoScriptError => {
                MY_SCRIPT
                    .key(key1)
                    .key(key2)
                    .arg(&arg1)
                    .arg(&arg2)
                    .invoke_async(&mut connection)
                    .await
                    .expect("Failed to load script")
            }
            _ => {
                error!("Unhandled Redis error");
                ...
            }
        }
    }
}
```

In my case, the script runs last, so I guess it will also fail last, making this OK?

**Comment by nihohit** [comment]
@sondrelg Yes, if no other call in the pipeline is a call to invoke script, then it makes sense to assume that the NoScriptError is from this call.
